src/Plugin/ContentTypePlugin.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Http\Client\Common\Plugin;
+
+use Http\Client\Common\Plugin;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Allow to set the correct content type header on the request automatically only if it is not set .
+ *
+ * @author Karim Pinchon <karim.pinchon@gmail.com>
+ */
+final class ContentTypePlugin implements Plugin
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        if (!$request->hasHeader('Content-Type')) {
+            $stream = $request->getBody();
+
+            if ($this->isJson($stream)) {
+                $request = $request->withHeader('Content-Type', 'application/json');
+
+                return $next($request);
+            }
+
+            if ($this->isXml($stream)) {
+                $request = $request->withHeader('Content-Type', 'application/xml');
+
+                return $next($request);
+            }
```

Comment:
Some content type that can be added to check (but not necessarly in this PR) :

 * Image types by using a library image (should be optional IMO)
 * Binary type by checking if there is a null character (\0) in the stream (that's how git does to detect binary vs text)
 * Text type (inverse of the binary type, imo the default content type we want to add to the stream)